### DISCUSSION TITLE
TLS 1.3 and schannel on Windows 10 21H1

### DISCUSSION THREAD

**Discussion by gvollant** [curl/curl#6958] [General]
I installed Windows 10 21H1 and tried idea from

https://github.com/curl/curl/issues/4918
(CURL_SSLVERSION_MAX_TLSv1_3  is now documented).

But, using:
    ssl_version_max = CURL_SSLVERSION_MAX_TLSv1_3;

and

    case CURL_SSLVERSION_TLSv1_3:
      schannel_cred->grbitEnabledProtocols |= SP_PROT_TLS1_3_CLIENT;
      break;

I've still error "SEC_E_ALGORITHM_MISMATCH (0x80090331)"

How run the TLS 1.3 support in Windows 10 21H1  ?

**Comment by gvollant** [comment]
I tried modification for TLS 1.3 support.
Tried with both Windows 10 version 21H1 19063.928 and beta Windows server 2022  - 21H2 Build 20339.2

I always has error
schannel: AcquireCredentialsHandle failed: SEC_E_ALGORITHM_MISMATCH (0x80090331) - The client and server cannot communicate, because they do not possess a common algorithm.

On Windows 2022 21H2, both Internet explorer and a WinInet application are able to get https://tls13.akamai.io/ 
Akamai server say TLS1.3 is used with TLS_AES_256_GCM_SHA384

@jay an idea?

```
 lib/vtls/schannel.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/lib/vtls/schannel.c b/lib/vtls/schannel.c
index f12f9a012..35c23d59b 100644
--- a/lib/vtls/schannel.c
+++ b/lib/vtls/schannel.c
@@ -117,6 +117,10 @@
 #define SP_PROT_TLS1_2_CLIENT           0x00000800
 #endif
 
+#ifndef SP_PROT_TLS1_3_CLIENT
+#define SP_PROT_TLS1_3_CLIENT           0x00002000
+#endif
+
 #ifndef SCH_USE_STRONG_CRYPTO
 #define SCH_USE_STRONG_CRYPTO           0x00400000
 #endif
@@ -177,7 +181,7 @@ set_ssl_version_min_max(SCHANNEL_CRED *schannel_cred, struct Curl_easy *data,
   switch(ssl_version_max) {
   case CURL_SSLVERSION_MAX_NONE:
   case CURL_SSLVERSION_MAX_DEFAULT:
-    ssl_version_max = CURL_SSLVERSION_MAX_TLSv1_2;
+    ssl_version_max = CURL_SSLVERSION_MAX_TLSv1_3;
     break;
   }
   for(; i <= (ssl_version_max >> 16); ++i) {
@@ -192,8 +196,8 @@ set_ssl_version_min_max(SCHANNEL_CRED *schannel_cred, struct Curl_easy *data,
       schannel_cred->grbitEnabledProtocols |= SP_PROT_TLS1_2_CLIENT;
       break;
     case CURL_SSLVERSION_TLSv1_3:
-      failf(data, "schannel: TLS 1.3 is not yet supported");
-      return CURLE_SSL_CONNECT_ERROR;
+      schannel_cred->grbitEnabledProtocols |= SP_PROT_TLS1_3_CLIENT;
+      break;
     }
   }
   return CURLE_OK;
```

**Comment by gvollant** [comment]
I made a branch on my curl fork : https://github.com/gvollant/curl/tree/gv_schannel_tls13

**Comment by bagder** [comment]
What's the intention of posting this as a discussion? Basically nobody will read it and if you have a change you want to propose you should rather do it as a pull request!

**Reply by gvollant** [reply]
I can made a pull request with my draft

**Reply by gvollant** [reply]
I created https://github.com/curl/curl/pull/7262
there was a discussion also on
https://curl.se/mail/lib-2021-04/0100.html

**Comment by AjayGitUser** [comment]
Hi,
I am still facing this issue with Windows-11,
posted my query below: 
https://stackoverflow.com/questions/70767482/has-anyone-tried-using-tls-1-3-using-schannel-with-windows-11-21h2

Please suggest if you got any solution for this error,
FYI: I am not using CURL Lib, It's with SChannel.

**Reply by bagder** [reply]
Please, this discussion is for curl use cases only.

**Reply by gvollant** [reply]
And go to https://github.com/curl/curl/pull/7784 focurl + schannel + tls 1.3
